lib/membrane_rtc_engine/engine.ex

```diff
@@ -994,80 +959,73 @@ defmodule Membrane.RTC.Engine do
-      Enum.map(endpoints_handling_raw_format, fn endpoint_id ->
-        prepare_track_to_endpoint_links(track_id, endpoint_id, :raw_format_tee)
-      end)
+    cond do
+      track == nil ->
+        {:error, :invalid_track_id}
 
-    links_to_remote_format_endpoints =
-      Enum.map(endpoints_handling_remote_format, fn endpoint_id ->
-        prepare_track_to_endpoint_links(track_id, endpoint_id, :tee)
-      end)
+      subscription.format not in track.format ->
+        {:error, :invalid_format}
 
-    links =
-      endpoint_to_tee_links ++
-        raw_format_links ++ links_to_raw_format_endpoints ++ links_to_remote_format_endpoints
+      # TODO maybe simulcast_encodings should be always a list
+      default_simulcast_encoding != nil and track.simulcast_encodings != nil and
+          default_simulcast_encoding not in track.simulcast_encodings ->
+        {:error, :invalid_default_simulcast_encoding}
 
-    state = %{state | waiting_for_linking: waiting_for_linking}
-    {links, state}
+      true ->
+        :ok
+    end
   end
 
-  defp link_outbound_tracks(tracks_formats, endpoint_id, ctx, state) do
-    {tracks_to_link, tracks_not_to_link} =
-      Enum.split_with(tracks_formats, fn {track_id, _format} ->
-        Map.has_key?(ctx.children, {:tee, track_id})
-      end)
+  defp try_fulfill_subscription(subscription, ctx, state) do
+    # if tee for this track is already spawned, fulfill subscription
+    # otherwise, save subscription as pending, we will fulfill it
+    # when tee appears
+    if Map.has_key?(ctx.children, {:tee, subscription.track_id}) do
+      fulfill_subscription(subscription, ctx, state)
+    else
+      state = update_in(state, [:pending_subscriptions], &[subscription | &1])
+      {[], state}
+    end
+  end
 
-    links =
-      Enum.flat_map(tracks_to_link, fn
-        {track_id, :raw} ->
-          if Map.has_key?(ctx.children, {:raw_format_tee, track_id}) do
-            []
-          else
-            prepare_raw_format_links(track_id, state)
-          end ++
-            prepare_track_to_endpoint_links(track_id, endpoint_id, :raw_format_tee)
-
-        {track_id, _remote_format} ->
-          prepare_track_to_endpoint_links(track_id, endpoint_id, :tee)
-      end)
+  defp fulfill_subscription(%Subscription{format: :raw} = s, ctx, state) do
+    raw_format_links =
+      if Map.has_key?(ctx.children, {:raw_format_tee, s.track_id}) do
+        []
+      else
+        prepare_raw_format_links(s.track_id, state)
+      end
 
-    waiting_for_linking = MapSet.new(tracks_not_to_link, fn {track_id, _format} -> track_id end)
+    {links, state} = do_fulfill_subscription(s, :raw_format_tee, state)
 
-    state =
-      update_in(
-        state,
-        [:waiting_for_linking, endpoint_id],
-        &MapSet.union(&1, waiting_for_linking)
-      )
+    {raw_format_links ++ links, state}
+  end
 
+  defp fulfill_subscription(%Subscription{format: _remote_format} = s, _ctx, state) do
```

Comment:
```suggestion
  defp fulfill_subscription(%Subscription{format: _remote_format} = subscription, _ctx, state) do
```

I'd avoid shortcuts and one-letter names